### PR TITLE
non-strict export with dynamic shapes

### PR DESCRIPTION
**Pull request by avikchaudhuri**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115387

Differential Revision: [D51959229](https://our.internmc.facebook.com/intern/diff/D51959229/)

### CODE REVIEW

File: torch/_export/__init__.py [hunk 1]
```diff
@@ -792,15 +792,21 @@ def strip_root(x):
...
         assert out_spec is not None
         return ExportedProgram(
             ep_non_strict.gm,
             ep_non_strict.gm.graph,
             ep_non_strict.sig,
             _get_params_buffers(f),
-            {},
-            [],
+            range_constraints,
+            equality_constraints,
```
Comment (zhxchen17) [inline torch/_export/__init__.py:809]: not blocking but I feel maybe we could stop returning equality constraints as they will be removed and no one actively use them?

**Review by zhxchen17 — COMMENTED** [review]
Looks good to me overall, will let @tugsbayasgalan comment on dynamic shape part

File: torch/_export/__init__.py (hunk 1 above)
Comment (avikchaudhuri) [reply torch/_export/__init__.py:809]: I want to preserve behavior for now to minimize blast radius. I can zero out all equality constraints in a different PR (I was under the impression this was already done...)

File: torch/_export/non_strict_utils.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+def make_fake_inputs(nn_module, args, constraints):
+    tensor_constraints: Dict[int, Dict[int, Constraint]] = defaultdict(dict)
+    for constraint in constraints:
+        tensor_constraints[constraint.t_id][constraint.dim] = constraint
+        if constraint.shared is not None:
+            tensor_constraints[constraint.shared.t_id][constraint.shared.dim] = constraint
+
+    code = nn_module.forward.__code__
+    co_fields = {
```
Comment (tugsbayasgalan) [inline torch/_export/non_strict_utils.py:83]: Just curious: why do we need these in the shape_env

File: torch/_export/non_strict_utils.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        "co_filename": code.co_filename,
+        "co_firstlineno": code.co_firstlineno,
+    }
+    with FakeTensorMode(
+        shape_env=ShapeEnv(tracked_fakes=[], co_fields=co_fields)
+    ) as fake_mode:
+        params = inspect.signature(nn_module.forward).parameters
+        dim_name_to_sources: Dict[str, List[Source]] = defaultdict(list)
+        fake_args = tuple(
+            fake_tree(
```
Comment (tugsbayasgalan) [inline torch/_export/non_strict_utils.py:94]: nit: Why can't we use pytree here?

File: torch/_export/non_strict_utils.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    )
+    shape_env.produce_guards(
+        placeholders,
+        sources,
+        constraint_inputs=constraint_inputs,
+        equalities_inputs=equalities_inputs,
+        ignore_static=False,
+    )
+    shape_env.frozen = True
+    shape_env.dim_constraints.solve()
```
Comment (tugsbayasgalan) [inline torch/_export/non_strict_utils.py:123]: Should we also move the constraint discovery too?

**Review by tugsbayasgalan — COMMENTED** [review]
Awesome :) Just have few questions

File: torch/_export/non_strict_utils.py (hunk 4 above)
Comment (avikchaudhuri) [reply torch/_export/non_strict_utils.py:123]: Yes I will, separately. It should follow the same pattern as here, but don't want to deal with potential strict test failures in this PR.

File: torch/_export/non_strict_utils.py (hunk 3 above)
Comment (avikchaudhuri) [reply torch/_export/non_strict_utils.py:94]: Yeah that would be nice, but unfortunately cannot. We're building sources too as we descend. (I can check if there's a pytree util that can do additional bookkeeping as we descend...)

File: torch/_export/non_strict_utils.py (hunk 2 above)
Comment (avikchaudhuri) [reply torch/_export/non_strict_utils.py:83]: I wanted the shape env to be as much the same as what's produced by strict, and I noticed that this is populated in strict mode.

File: torch/_export/non_strict_utils.py (hunk 4 above)
Comment (avikchaudhuri) [reply torch/_export/non_strict_utils.py:132]: Wait, I'm not sure what you mean...I was assuming you mean, can we move around steps in strict mode to more closely reflect what's going on here.

As for constraint discovery, well that is essentially `.solve()` + pretty-printing the suggestions. And yes, I forgot to add that. There's actually some non-trivial logic around constraint violation errors that we need to replicate from `eval_frame`, including turning them into user errors, catching them, adding some message related to constraint discovery, etc.

Comment (tugsbayasgalan) [reply torch/_export/non_strict_utils.py:132]: Yep I was referring to .solve() + pretty printing the suggestions.

**Review by tugsbayasgalan — APPROVED** [review]

**Pull request closed without merging**